### ISSUE TITLE
Request: staging through REST API (or WebDAV doors)

### ISSUE THREAD

**Issue by onnozweers** [dCache/dcache#4349]
Dear dCache developers,

We would very much welcome a possibility to stage files through the REST API, or, if that is not possible or desirable, through WebDAV doors. The reason is that we're developing a high throughput cluster that will not use X509 authentication. AFAIK staging is possible through NFS dot files, but NFS will (probably) not be used. Additional wishes: we'd like bulk/bundle stage requests to be possible, and we'd like basic authentication as well as macaroon authentication to be possible.

AFAIK staging with username/password should be possible by reading 1 byte of a tape file through WebDAV. However, this is not very elegant.

Testing with a 5.0 snapshot API, we've already seen the methods to get the locality and checksums of files. We think this would be a great addition, especially for those users who are not familiar with grid tools and grid authentication.

Kind regards,
Onno

**Comment by vingar** [comment]
Some examples with curl:
Bring online:
```curl -L  --capath /etc/grid-security/certificates --cert /tmp/x509up_u`id -u` --cacert /tmp/x509up_u`id -u` --key  /tmp/x509up_u`id -u`  -X POST -H "Accept: application/json" -H "Content-Type: application/json" https://hostname:3880/api/v1/namespace/file --data '{"action" : "qos", "target" : "disk+tape"}'
{"status":"success"}```

To query the status:
```$ curl  -k -X GET " https://hostname:3880/api/v1/namespace/file?locality=true"
{
  "fileMimeType" : "application/gzip",
  "fileLocality" : "ONLINE_AND_NEARLINE",
  "pnfsId" : "0000BB50C48197C746069E74EDE9FB0BE73F",
  "fileType" : "REGULAR",
  "nlink" : 1,
  "mtime" : 1539897746207,
  "size" : 697876196,
  "creationTime" : 1539897202182
}```

I've written / generate some python clients for my own usage if you are interested I can share them.

**Comment by onnozweers** [comment]
@paulmillar Thanks, we had overlooked this! I guess we need to get used to `QoS` as a synonym for staging and restoring. We'll test this out soon. I've not noticed support for bulk requests; we'll see how single file requests perform. Does the API support macaroon authentication?

**Comment by onnozweers** [comment]
@vingar We'd be very interested!

**Comment by paulmillar** [comment]
Yes, the longer-term plan is to replace the DISK / TAPE dichotomy with a more general concept of QoS.  This is the first (baby) steps in that direction.

We haven't implemented support for bulk operation ... yet.  This work was driven by supporting a CDMI interface, where these QoS concepts may be expressed in a standard fashion (since CDMI is a standard).  Currently, CDMI doesn't support bulk operations, so this wasn't implemented.

With people increasingly looking at alternatives to SRM, I suspect some kind of bulk operation will be needed, so this will likely change in the future.

**Comment by onnozweers** [comment]
I've noticed that there is an example specified in the code at https://github.com/dCache/dcache/blob/master/modules/dcache-frontend/src/main/java/org/dcache/restful/resources/namespace/FileResources.java:

    examples = @Example({
        @ExampleProperty("{\n"
            + "    \"action\" : \"mv\""
            + "    \"destination\" : \"../foo\""
            + "}")}))

Yet in the API I can only see `"string"` in the example value. Is that a bug? Or am I looking at the wrong piece of code?

**Comment by vingar** [comment]
@onnozweers 

My python clients are there: https://github.com/vingar/dcacheclient also available on pypi https://pypi.org/project/dcacheclient/.

Some info relevant for this issue:

Command line:
```
dcache-admin  namespace cmrResources  --path /test/tape/file.ff1eb8f8-d845-11e8-a22c-acde48001122 --body '{"action": "qos", "target": "disk+tape"}'
```
Alias:
```
dcache-admin namespace bring-online  --path /test/tape/file.ff1eb8f8-d845-11e8-a22c-acde48001122
```
With the python APIs:
```
>>> from dcacheclient.client import Client
>>> dcache = Client(url=...)

>>> dcache.namespace.bringonline(path='/test/tape/file.ff1eb8f8-d845-11e8-a22c-acde48001122')
```

```
>>> from dcacheclient.client import Client
>>> dcache = Client(url=...)
>>> dcache.namespace.cmr_resources(path='/test/tape/file.ff1eb8f8-d845-11e8-a22c-acde48001122', body='{"action": "qos", "target": "disk+tape"}')
# Alias
>>> dcache.namespace.bringonline(path='/test/tape/file.ff1eb8f8-d845-11e8-a22c-acde48001122')
..
```

**Comment by DmitryLitvintsev** [comment]
Just a word of warning. RESTful API call to change qos from 'tape' to 'disk+tape' will result in permanent pin of file replica on the pool where that files gets restored to. So while you do get 
"a possibility to stage files through the REST API" be aware of this consequence. To unpin the file you need to run the opposite QOS transition:

```
curl -L --capath /etc/grid-security/certificates --cert /tmp/x509up_u`id -u` --cacert /tmp/x509up_u`id -u` --key /tmp/x509up_u`id -u` -X POST -H "Accept: application/json" -H "Content-Type: application/json" https://hostname:3880/api/v1/namespace/file --data '{"action" : "qos", "target" : "tape"}'
```

IOW the RESTFul API action is equivalent to  `bring-online` with unlimited lifetime.

**Comment by onnozweers** [comment]
Thanks Dmitry, that is very important for us to know.

I have a question about this. Let's say there are two users working with the same dataset. Both change the QoS of a file to disk. But when one of them is done and changes the QoS back to tape, it would interfere with the work of the other user, who suddenly finds the file not online anymore. Would you have a suggestion on how to handle that? Or is it up to the users to coordinate their work?

Kind regards,
Onno

**Comment by onnozweers** [comment]
Hi Paul,

Thanks for the additional explanation. Good to know that files are not actively & instantly removed to meet the QoS, but that the usual caching applies.

**Comment by onnozweers** [comment]
One more question: does the API support macaroon authentication? IOW, can we give people macaroons that they can stage with?

**Comment by paulmillar** [comment]
Yes, macaroons will work with the REST API.
